### PR TITLE
Use Int over NSNumber in logit filters and related extensions

### PR DESCRIPTION
**Pull request by naykutguven**
This pull request refactors the handling of multi-dimensional array indices throughout the codebase, standardizing on the use of `[Int]` instead of `[NSNumber]` for improved type safety and performance. The changes affect both the main implementation and the test suite, updating method signatures, internal logic, and usage sites. Deprecated methods are introduced for backward compatibility.

**Core API and Implementation Updates:**

* Updated the `MLMultiArray` extension methods (`linearOffset(for:)` and `fill(indexes:with:)`) to use `[Int]` instead of `[NSNumber]`, and added deprecated overloads to maintain backward compatibility.
* Refactored all logits filter classes (`SuppressTokensFilter`, `SuppressBlankFilter`, `TimestampRulesFilter`, `LanguageLogitsFilter`) to use `[[Int]]` for index arrays instead of `[[NSNumber]]`.

**Test Suite Adjustments:**

* Updated test utility methods and test cases to use `[Int]` and `[[Int]]` for indices, aligning with the new API.

These changes modernize the codebase, simplify type handling, and reduce unnecessary type conversions, while ensuring backward compatibility for existing code.

### CODE REVIEW

**Comment by naykutguven** [comment]
The benchmark result of the script below:

```
newLinearOffset([Int]): 52.43 ms (checksum: 682460000000)
oldLinearOffset([NSNumber]): 513.97 ms (checksum: 682460000000)
bridgedLinearOffset([NSNumber] -> [Int]): 936.11 ms (checksum: 682460000000)
fill([[Int]]): 27.27 ms (checksum: 268973232000)
fill([[NSNumber]]) old path: 300.60 ms (checksum: 268973232000)
fill([[NSNumber]]) bridged path: 538.53 ms (checksum: 268973232000)
approx index storage bytes [[Int]]: 1656512
approx index storage bytes [[NSNumber]] refs only: 1656512
```

You can use the script below to check the benchmark yourself:

```swift
import Foundation

@inline(__always)
func oldLinearOffset(index: [NSNumber], strides: [Int]) -> Int {
    var linearOffset = 0
    for (dimension, stride) in zip(index, strides) {
        linearOffset += dimension.intValue * stride
    }
    return linearOffset
}

@inline(__always)
func newLinearOffset(index: [Int], strides: [Int]) -> Int {
    var linearOffset = 0
    for (dimension, stride) in zip(index, strides) {
        linearOffset += dimension * stride
    }
    return linearOffset
}

@inline(__always)
func bridgedLinearOffset(index: [NSNumber], strides: [Int]) -> Int {
    newLinearOffset(index: index.map(\.intValue), strides: strides)
}

func measure(_ name: String, _ body: () -> Int) {
    let start = DispatchTime.now().uptimeNanoseconds
    let checksum = body()
    let end = DispatchTime.now().uptimeNanoseconds
    let ms = Double(end - start) / 1_000_000.0
    let msString = String(format: "%.2f", ms)
    print("\(name): \(msString) ms (checksum: \(checksum))")
}

let strides = [51865, 51865, 1]
let intIndex = [0, 0, 34123]
let nsIndex: [NSNumber] = [0, 0, 34123]
let iterations = 20_000_000

measure("newLinearOffset([Int])") {
    var sum = 0
    for _ in 0..<iterations {
        sum &+= newLinearOffset(index: intIndex, strides: strides)
    }
    return sum
}

measure("oldLinearOffset([NSNumber])") {
    var sum = 0
    for _ in 0..<iterations {
        sum &+= oldLinearOffset(index: nsIndex, strides: strides)
    }
    return sum
}

measure("bridgedLinearOffset([NSNumber] -> [Int])") {
    var sum = 0
    for _ in 0..<iterations {
        sum &+= bridgedLinearOffset(index: nsIndex, strides: strides)
    }
    return sum
}

let logitsDim = 51_865
let languageTokenCount = 99
let allLanguageTokens = Set((0..<languageTokenCount).map { $0 * 20 })

let intIndexes: [[Int]] = (0..<logitsDim).compactMap { i in
    allLanguageTokens.contains(i) ? nil : [0, 0, i]
}

let nsIndexes: [[NSNumber]] = (0..<logitsDim).compactMap { i in
    allLanguageTokens.contains(i) ? nil : [0, 0, NSNumber(value: i)]
}

let elementCount = logitsDim
let fillIterations = 200

func fillInt(indexes: [[Int]], storage: inout [Float]) -> Int {
    var checksum = 0
    for _ in 0..<fillIterations {
        for index in indexes {
            let offset = newLinearOffset(index: index, strides: strides)
            storage[offset % elementCount] = -.infinity
            checksum &+= offset
        }
    }
    return checksum
}

func fillNSNumber(indexes: [[NSNumber]], storage: inout [Float]) -> Int {
    var checksum = 0
    for _ in 0..<fillIterations {
        for index in indexes {
            let offset = oldLinearOffset(index: index, strides: strides)
            storage[offset % elementCount] = -.infinity
            checksum &+= offset
        }
    }
    return checksum
}

func fillBridged(indexes: [[NSNumber]], storage: inout [Float]) -> Int {
    var checksum = 0
    for _ in 0..<fillIterations {
        for index in indexes {
            let offset = bridgedLinearOffset(index: index, strides: strides)
            storage[offset % elementCount] = -.infinity
            checksum &+= offset
        }
    }
    return checksum
}

var storageA = Array<Float>(repeating: 0, count: elementCount)
var storageB = Array<Float>(repeating: 0, count: elementCount)
var storageC = Array<Float>(repeating: 0, count: elementCount)

measure("fill([[Int]])") {
    fillInt(indexes: intIndexes, storage: &storageA)
}

measure("fill([[NSNumber]]) old path") {
    fillNSNumber(indexes: nsIndexes, storage: &storageB)
}

measure("fill([[NSNumber]]) bridged path") {
    fillBridged(indexes: nsIndexes, storage: &storageC)
}

func approxBytesForIntIndexes(_ indexes: [[Int]]) -> Int {
    let outerStorage = indexes.count * MemoryLayout<[Int]>.stride
    let innerBuffers = indexes.reduce(0) { $0 + ($1.capacity * MemoryLayout<Int>.stride) }
    return outerStorage + innerBuffers
}

func approxBytesForNSNumberIndexes(_ indexes: [[NSNumber]]) -> Int {
    let outerStorage = indexes.count * MemoryLayout<[NSNumber]>.stride
    let innerBuffers = indexes.reduce(0) { $0 + ($1.capacity * MemoryLayout<NSNumber>.stride) }
    return outerStorage + innerBuffers
}

let intBytes = approxBytesForIntIndexes(intIndexes)
let nsBytes = approxBytesForNSNumberIndexes(nsIndexes)
print("approx index storage bytes [[Int]]: \(intBytes)")
print("approx index storage bytes [[NSNumber]] refs only: \(nsBytes)")
```
